src/protocol/taiko_alethia/DelayedInclusionStore.sol

```diff
@@ -0,0 +1,100 @@
+    /// @param dueInclusion The delayed publication including
+    /// the blob reference hash and due timestamp
+    /// @param sender The address that stored the delayed publication
+    event DelayedInclusionStored(address indexed sender, DueInclusion dueInclusion);
+
+    /// @notice Emitted when a list of delayed inclusions are processed by the inbox
+    /// @param inclusionsList list of inclusions
+    event DelayedInclusionProcessed(Inclusion[] inclusionsList);
+
+    // Append-only queue for delayed inclusions
+    DueInclusion[] private delayedInclusions;
+
+    // Pointer to the first unprocessed element
+    uint256 private head;
+
+    /// @notice The minimum amount of time a delayed publication needs to
+    /// wait in the queue to be included expressed in seconds
+    uint256 public immutable inclusionDelay;
+
+    address public immutable inbox;
+
+    IBlobRefRegistry public immutable blobRefRegistry;
+
+    /// @param _inclusionDelay The delay before next set of inclusions can be processed.
+    /// @param _blobRefRegistry The address of the blob reference registry contract.
+    /// @param _taikoInbox The address of the Taiko's inbox contract responsible for processing inclusions.
+    constructor(uint256 _inclusionDelay, address _blobRefRegistry, address _taikoInbox) {
+        inclusionDelay = _inclusionDelay;
+        inbox = _taikoInbox;
+        blobRefRegistry = IBlobRefRegistry(_blobRefRegistry);
+    }
+
+    /// @inheritdoc IDelayedInclusionStore
+    /// @dev Stores the blob reference as a DueInclusion
+    function publishDelayed(uint256[] calldata blobIndices) external {
+        bytes32 refHash = keccak256(abi.encode(blobRefRegistry.getRef(blobIndices)));
+        DueInclusion memory dueInclusion = DueInclusion(refHash, block.timestamp + inclusionDelay);
+        delayedInclusions.push(dueInclusion);
+        emit DelayedInclusionStored(msg.sender, dueInclusion);
+    }
+
+    /// @inheritdoc IDelayedInclusionStore
+    /// @dev Only returns inclusions if the delay period has passed
+    /// otherwise returns an empty array.
+    /// @dev Can only be called by the inbox contract.
+    function processDueInclusions() external returns (Inclusion[] memory) {
+        uint256 len = delayedInclusions.length;
+        if (head >= len) {
+            return new Inclusion[](0);
+        }
+
+        uint256 l = head;
+        uint256 r = len;
+        uint256 timestamp = block.timestamp;
+        while (l < r) {
+            uint256 mid = (l + r) >> 1;
+            if (delayedInclusions[mid].due <= timestamp) {
+                // If inclusion at mid is due, search to the right.
+                l = mid + 1;
+            } else {
+                r = mid;
+            }
+        }
+
+        // l now is the first index where the inclusion is not yet due.
+        uint256 count = l - head;
+        if (count == 0) {
+            return new Inclusion[](0);
+        }
```

Comment:
I would abstract this into an internal function (e.g. `_countDelayedInclusions`). In my opinion, if anyone wants to rely on a mapping instead they should be able to do it by overriding `_countDelayedInclusions`

Perhaps:

```solidity
function _countDelayedInclusions() internal virtual returns (uint256 last, uint256 count) {
        uint256 l = head;
        uint256 r = len;
        uint256 timestamp = block.timestamp;
        while (l < r) {
            uint256 mid = (l + r) >> 1;
            if (delayedInclusions[mid].due <= timestamp) {
                // If inclusion at mid is due, search to the right.
                l = mid + 1;
            } else {
                r = mid;
            }
        }

        return (l, l - head);
}
```